src/geo/projection/globe_util.js

```diff
@@ -409,103 +398,139 @@ export function globePoleMatrixForTile(z: number, x: number, tr: Transform): Flo
-    gridSegments: SegmentVector;
+    _gridBuffer: VertexBuffer;
+    _gridIndexBuffer: IndexBuffer;
+    _gridSegments: Array<SegmentVector>;
 
     atmosphereVertexBuffer: VertexBuffer;
     atmosphereIndexBuffer: IndexBuffer;
     atmosphereSegments: SegmentVector;
 
-    wireframeIndexBuffer: IndexBuffer;
-    wireframeSegments: SegmentVector;
+    _wireframeIndexBuffer: IndexBuffer;
+    _wireframeSegments: Array<SegmentVector>;
 
     constructor(context: Context) {
+        assert(GLOBE_VERTEX_GRID_SIZE_LOD_TABLE.length === GLOBE_ZOOM_THRESHOLD_MIN);
         this._createGrid(context);
         this._createPoles(context);
         this._createAtmosphere(context);
     }
 
     destroy() {
-        this.poleIndexBuffer.destroy();
-        this.gridIndexBuffer.destroy();
-        this.poleNorthVertexBuffer.destroy();
-        this.poleSouthVertexBuffer.destroy();
-        for (const segments of this.poleSegments) segments.destroy();
-        this.gridSegments.destroy();
+        this._poleIndexBuffer.destroy();
+        this._gridBuffer.destroy();
+        this._gridIndexBuffer.destroy();
+        this._poleNorthVertexBuffer.destroy();
+        this._poleSouthVertexBuffer.destroy();
+        for (const segments of this._poleSegments) segments.destroy();
+        for (const segments of this._gridSegments) segments.destroy();
         this.atmosphereVertexBuffer.destroy();
         this.atmosphereIndexBuffer.destroy();
         this.atmosphereSegments.destroy();
 
-        if (this.wireframeIndexBuffer) {
-            this.wireframeIndexBuffer.destroy();
-            this.wireframeSegments.destroy();
+        if (this._wireframeIndexBuffer) {
+            this._wireframeIndexBuffer.destroy();
+            for (const segments of this._wireframeSegments) segments.destroy();
         }
     }
 
     _createGrid(context: Context) {
+        // All tiles use the same vertex and index buffer. Different segments view into the index buffer.
+        // This allows picking different "resolution" for a given tile based on its zoom level.
+        const gridVertices = new PosArray();
         const gridIndices = new TriangleIndexArray();
-        const quadExt = GLOBE_VERTEX_GRID_SIZE;
-        const vertexExt = quadExt + 1;
-
-        for (let j = 0; j < quadExt; j++) {
-            for (let i = 0; i < quadExt; i++) {
-                const index = j * vertexExt + i;
-                gridIndices.emplaceBack(index + 1, index, index + vertexExt);
-                gridIndices.emplaceBack(index + vertexExt, index + vertexExt + 1, index + 1);
+
+        const maxVertices = MAX_VERTEX_GRID_SIZE + 1;
+
+        this._gridSegments = [];
+
+        for (let j = 0; j < maxVertices; j++)
+            for (let i = 0; i < maxVertices; i++)
+                gridVertices.emplaceBack(i, j);
+
+        for (let z = 0, primitiveOffset = 0; z < GLOBE_ZOOM_THRESHOLD_MIN; z++) {
+            const quadExt = GLOBE_VERTEX_GRID_SIZE_LOD_TABLE[z];
```

Comment:
@mpulkki-mapbox I was aware of the issue of tiles not being watertight but this problem could still manifest itself before this change and with POT values although with POT it would be harder to notice it. My goal was to find a good set of values for the LOD table to mitigate this issue. Adding tile latitude to the equation would make this even more challenging cc @karimnaaji. I will investigate this further.
I will also investigate the issue with the wireframe thanks for the catch!

Reply:
@mpulkki-mapbox regarding the code snippet currently the LOD table is chosen to have the same size as the minimum globe zoom threshold. This is enforced via an assertion so that the ranges are aligned.

Reply:
@akoylasar, good point, the issue with borders not being watertight is not a regression :)

Reply:
@mpulkki-mapbox unfortunately though the wireframe issue is caused by precision loss. This in fact can become seriously noticeable when terrain is enabled as seen below:
![terrain_globeview_precisionloss](https://user-images.githubusercontent.com/67509069/157437219-4bdf4c8c-8828-40d3-b8c9-346e3672bc39.png)
I will try to think if there's a way to preserve precision because this in fact could be a deal breaker.

Reply:
I don't think this is a precision issue. Looks like that the elevation sampling is not setup correctly after the latest changes

Reply:
@mpulkki-mapbox I haven't modified the elevation code when it comes to rendering the non-pole tiles. Perhaps I have messed up something and have overlooked it. Could you please point me to where you think the issue is?

Reply:
Ha! I found the issue and you were right it wasn't due to precision issues so I think I jumped to conclusion somewhat quickly. Anyways the issue was this:
```
vec4 globe = u_globe_matrix * vec4(globe_pos + up_vector.xyz * height, 1.0);
```
previously `globe_pos` was of course normalized but this is no longer the case since we calculate it in the shader. I simply overlooked the fact that the `up_vector` is also in the same space. What should have been done however is the following:
```
vec4 vv = u_normalized_globe_matrix * vec4(up_vector.xyz * height, 0.0);
vec4 globe = u_globe_matrix * vec4(globe_pos, 1.0);
globe = globe + vv;
```
It took me lots of debugging to figure notice this.. one of those cases! Sorry for the confusion and thanks again!

Reply:
Good catch! The shader uniform `up_vector` contains both a direction and a scale for converting elevation values (in meters) to whatever the tile coordinate space is. I think you could even simplify this a bit by just changing the value of `u_tile_up_scale` after invoking `terrain.setupElevationDraw` to a correct conversion value (meters-to-pixels).